qutebrowser/keyinput/modeman.py

```diff
@@ -247,6 +250,12 @@ def enter(self, mode, reason=None, only_if_normal=False):
                 return
             log.modes.debug("Overriding mode {}.".format(self.mode))
             self.left.emit(self.mode, mode, self._win_id)
+        if (mode in prompt_modes and self.mode in input_modes and
+                config.val.tabs.mode_on_change == 'restore'):
```

Comment:
I've thought about this some more (and found #3324 again) - I think this should be done independently from `tabs.mode_on_change`.

Reply:
Oh, I never saw that issue... interesting. So you want to undo these in here and create a different PR for issue #3324 ?

Reply:
In the future it'd be good to have different PRs for different futures, but since I already reviewed this and all, feel free to continue it as part of this PR.